homeassistant/components/growatt_server/config_flow.py

```diff
@@ -36,6 +40,14 @@ async def async_step_user(self, user_input=None):
         if not user_input:
             return self._async_show_user_form()
 
+        # Remove any accidental whitespace from the URL
+        user_input[CONF_URL] = user_input[CONF_URL].strip()
+
+        # If the URL doesn't end in a / add one (the python library expects a /)
+        if user_input[CONF_URL] != "" and user_input[CONF_URL][-1] != "/":
+            user_input[CONF_URL] = user_input[CONF_URL] + "/"
```

Comment:
In general, is it better to use a URL processing library instead of trying to string-process URLs. E.g., using `yarl`.

Reply:
I'll look into `yarl` this is the first time I've ever heard of it.

Reply:
Have removed yarl logic now since the URLs are all pre-formatted choices.